File: backend/Models/Booking.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace intergalactica.Models;

public class Booking
{
    [BsonId]
    [BsonRepresentation(BsonType.ObjectId)]
    public string Id { get; set; } = string.Empty;

    [BsonElement("source")] public ObjectId Source { get; set; } = new ObjectId(string.Empty);
    [BsonElement("destination")] public ObjectId Destination { get; set; } = new ObjectId(string.Empty);
    [BsonElement("source_dock")] public String SourceDock { get; set; } = string.Empty;
    [BsonElement("destination_dock")] public String DestinationDock { get; set; } = string.Empty;

    [BsonElement("departure")] public DateTime Departure { get; set; } = DateTime.MinValue;
    [BsonElement("arrival")] public DateTime Arrival { get; set; } = DateTime.MinValue;
    [BsonElement("return")] public bool Return { get; set; } = false;
    [BsonElement("departure_return")] public DateTime DepartureReturn { get; set; } = DateTime.MinValue;
    [BsonElement("arrival_return")] public DateTime ArrivalReturn { get; set; } = DateTime.MinValue;

    [BsonElement("user")] public ObjectId User { get; set; } = new ObjectId(string.Empty);
    [BsonElement("shuttle")] public ObjectId Shuttle { get; set; } = new ObjectId(string.Empty);
    [BsonElement("adults")] public int Adults { get; set; } = 0;
    [BsonElement("children")] public int Children { get; set; } = 0;

    [BsonElement("price")] public float Price { get; set; } = 0;
    [BsonElement("discount")] public float Discount { get; set; } = 0;
    [BsonElement("total")] public float Total { get; set; } = 0;
    [BsonElement("paid")] public bool Paid { get; set; } = false;


}

File: frontend2/lib/models/booking.dart
class Booking {
  String? id;
  String? source;
  String? destination;
  String? sourceDock;
  String? destinationDock;
  DateTime? departure;
  DateTime? arrival;
  bool? returnTrip;
  DateTime? departureReturn;
  DateTime? arrivalReturn;
  String? user;
  String? shuttle;
  int? adults;
  int? children;
  double? price;
  double? discount;
  double? total;
  bool? paid;

  Booking({
    this.id,
    this.source,
    this.destination,
    this.sourceDock,
    this.destinationDock,
    this.departure,
    this.arrival,
    this.returnTrip,
    this.departureReturn,
    this.arrivalReturn,
    this.user,
    this.shuttle,
    this.adults,
    this.children,
    this.price,
    this.discount,
    this.total,
    this.paid,
  });

  static Booking fromJson(jsonDecode) {
    return Booking(
      id: jsonDecode['id'],
      source: jsonDecode['source'],
      destination: jsonDecode['destination'],
      sourceDock: jsonDecode['sourceDock'],
      destinationDock: jsonDecode['destinationDock'],
      departure: DateTime.parse(jsonDecode['departure']),
      arrival: DateTime.parse(jsonDecode['arrival']),
      returnTrip: jsonDecode['returnTrip'],
      departureReturn: DateTime.parse(jsonDecode['departureReturn']),
      arrivalReturn: DateTime.parse(jsonDecode['arrivalReturn']),
      user: jsonDecode['user'],
      shuttle: jsonDecode['shuttle'],
      adults: jsonDecode['adults'],
      children: jsonDecode['children'],
      price: jsonDecode['price'],
      discount: jsonDecode['discount'],
      total: jsonDecode['total'],
      paid: jsonDecode['paid'],
    );
  }

  static Map<String, dynamic> toJson(Booking booking) {
    return {
      'source': booking.source,
      'destination': booking.destination,
      'sourceDock': booking.sourceDock,
      'destinationDock': booking.destinationDock,
      'departure': booking.departure?.toIso8601String(),
      'arrival': booking.arrival?.toIso8601String(),
      'returnTrip': booking.returnTrip,
      'departureReturn': booking.departureReturn?.toIso8601String(),
      'arrivalReturn': booking.arrivalReturn?.toIso8601String(),
      'user': booking.user,
      'shuttle': booking.shuttle,
      'adults': booking.adults,
      'children': booking.children,
      'price': booking.price,
      'discount': booking.discount,
      'total': booking.total,
      'paid': booking.paid,
    };
  }

  @override
  String toString() {
    return '\nBooking{id: $id, source: $source, destination: $destination, sourceDock: $sourceDock, destinationDock: $destinationDock, departure: $departure, arrival: $arrival, returnTrip: $returnTrip, departureReturn: $departureReturn, arrivalReturn: $arrivalReturn, user: $user, shuttle: $shuttle, adults: $adults, children: $children, price: $price, discount: $discount, total: $total, paid: $paid}\n';
  }
}
